File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/tracker/StateMachineInterface.java
package com.snowplowanalytics.snowplow.internal.tracker;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.event.Event;
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson;
import com.snowplowanalytics.snowplow.tracker.InspectableEvent;

import java.util.List;
import java.util.Map;

public interface StateMachineInterface {

    @NonNull
    List<String> subscribedEventSchemasForTransitions();

    @NonNull
    List<String> subscribedEventSchemasForEntitiesGeneration();

    @NonNull
    List<String> subscribedEventSchemasForPayloadUpdating();

    @Nullable
    State transition(@NonNull Event event, @Nullable State state);

    @Nullable
    List<SelfDescribingJson> entities(@NonNull InspectableEvent event, @Nullable State state);

    @Nullable
    Map<String, Object> payloadValues(@NonNull InspectableEvent event, @Nullable State state);
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/tracker/StateMachineInterface.kt
package com.snowplowanalytics.core.tracker

import com.snowplowanalytics.snowplow.event.Event
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson
import com.snowplowanalytics.snowplow.tracker.InspectableEvent

interface StateMachineInterface {
    val subscribedEventSchemasForTransitions: List<String>
    val subscribedEventSchemasForEntitiesGeneration: List<String>
    val subscribedEventSchemasForPayloadUpdating: List<String>
    fun transition(event: Event, state: State?): State?
    fun entities(event: InspectableEvent, state: State?): List<SelfDescribingJson>?
    fun payloadValues(event: InspectableEvent, state: State?): Map<String, Any>?
}
